py/objset.c

```diff
@@ -227,29 +227,45 @@ static mp_obj_t set_diff_update(size_t n_args, const mp_obj_t *args) {
 }
 static MP_DEFINE_CONST_FUN_OBJ_VAR(set_diff_update_obj, 1, set_diff_update);
 
-static mp_obj_t set_intersect_int(mp_obj_t self_in, mp_obj_t other, bool update) {
+static mp_obj_t set_intersect_int(size_t n_args, const mp_obj_t *args, bool update) {
     if (update) {
-        check_set(self_in);
+        check_set(args[0]);
     } else {
-        check_set_or_frozenset(self_in);
+        check_set_or_frozenset(args[0]);
     }
 
-    if (self_in == other) {
-        return update ? mp_const_none : set_copy(self_in);
+    if (n_args == 2 && args[0] == args[1]) {
+        return update ? mp_const_none : set_copy(args[0]);
     }
 
-    mp_obj_set_t *self = MP_OBJ_TO_PTR(self_in);
     mp_obj_set_t *out = MP_OBJ_TO_PTR(mp_obj_new_set(0, NULL));
 
-    mp_obj_t iter = mp_getiter(other, NULL);
-    mp_obj_t next;
-    while ((next = mp_iternext(iter)) != MP_OBJ_STOP_ITERATION) {
-        if (mp_set_lookup(&self->set, next, MP_MAP_LOOKUP)) {
-            set_add(MP_OBJ_FROM_PTR(out), next);
+    mp_obj_t self_iter = mp_getiter(args[0], NULL);
+    mp_obj_t self_next;
+    while ((self_next = mp_iternext(self_iter)) != MP_OBJ_STOP_ITERATION) {
+        bool skip = false;
+        for (size_t i = 1; i < n_args; i++) {
+            mp_obj_t iter = mp_getiter(args[i], NULL);
+            mp_obj_t next;
+            bool present = false;
+            while ((next = mp_iternext(iter)) != MP_OBJ_STOP_ITERATION) {
```

Comment:
This looks like it is doing a linear search now instead of `mp_set_lookup()` (which uses a hash-based search), so we would need to see some performance measurements to compare this to the previous implementation for large sets.

Reply:
I've taken a look at CPython and it seems CPython has a function for single intersection and a function for multi intersections (takes multiple iterables) which calculates `n-1` intersections. I guess code size increase to define a new function is more than this linear search BTW.

Reply:
We definitely don't want to reduce the performance of the existing hash-based approach.

Reply:
Hmm... What about keeping the existing while loop structure around the first argument, and checking any additional args inside the loop?

Semi-pseudocode:

```
self = args[0];
other = args[1];

 mp_obj_t iter = mp_getiter(other, NULL);
 mp_obj_t next;
 while ((next = mp_iternext(iter)) != MP_OBJ_STOP_ITERATION) {
        if (mp_set_lookup(&self->set, next, MP_MAP_LOOKUP)) {
            for (int i = 2; i < n_args; i++) {
                if (!mp_binary_op(MP_BINARY_OP_CONTAINS(args[i], next)) {
                    continue; // Item not in all arguments
               }
            }
            set_add(MP_OBJ_FROM_PTR(out), next);
```

This should perform roughly the same as the existing code with one arg (due to the set lookup), and multiple args should be faster than calling the existing function multiple times (as the binary "contains" can be faster if the arguments are large sets).